staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,619 @@
+	qs.DequeueWithChannelAsMuchAsPossible()
+}
+
+// FinishRequest is a callback that should be used when a previously dequeued packet
+// has completed it's service.  This callback updates imporatnt state in the
+// queueSetImpl
+func (qs *queueSetImpl) finishRequest(r *fq.Request) {
+
+	S := qs.clk.Since(r.StartTime).Seconds()
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	r.Queue.VirStart -= qs.estimatedServiceTime - S
+
+	// request has finished, remove from requests executing
+	r.Queue.RequestsExecuting--
+
+	// Logic to remove quiesced queues
+	// >= as QueueIdx=25 is out of bounds for desiredNumQueues=25 [0...24]
+	if r.Queue.Index >= qs.desiredNumQueues &&
+		len(r.Queue.Requests) == 0 &&
+		r.Queue.RequestsExecuting == 0 {
+		qs.queues = removeQueueAndUpdateIndexes(qs.queues, r.Queue.Index)
+
+		// decrement here to maintain the invariant that (qs.robinIdx+1) % numQueues
+		// is the index of the next queue after the one last dispatched from
+		if qs.robinIdx > -r.Queue.Index {
+			qs.robinIdx--
+		}
+
+		// At this point, if the qs is quiescing,
+		// has zero requests executing, and has zero requests enqueued
+		// then a call to the EmptyHandler should be forked.
+		if qs.emptyHandler != nil && qs.numRequestsEnqueued == 0 &&
+			qs.GetRequestsExecuting() == 0 {
+			qs.counter.Add(1)
+			go qs.emptyHandler.HandleEmpty()
+		}
+	}
+}
+
+// dequeue dequeues a packet from the queueSetImpl
+func (qs *queueSetImpl) dequeue() (*fq.Request, bool) {
+
+	queue := qs.selectQueue()
+
+	if queue == nil {
+		return nil, false
+	}
+	packet, ok := queue.Dequeue()
+
+	if ok {
+		// When a request is dequeued for service -> qs.VirStart += G
+		queue.VirStart += qs.estimatedServiceTime
+
+		packet.StartTime = qs.clk.Now()
+
+		// request dequeued, service has started
+		queue.RequestsExecuting++
+		metrics.UpdateFlowControlRequestsExecuting(qs.name, queue.RequestsExecuting)
+	} else {
+		// TODO(aaron-prindle) verify this statement is needed...
+		return nil, false
+	}
+	qs.numRequestsEnqueued--
+	return packet, ok
+}
+
+// Dequeue dequeues a packet from the queueSetImpl
+func (qs *queueSetImpl) Dequeue() (*fq.Request, bool) {
+	qs.lockAndSyncTime()
+	defer qs.lock.Unlock()
+	return qs.dequeue()
+}
+
+// isEmpty is a convenience method that returns 'true' when all of the queues
+// in an queueSetImpl have no packets (and is "empty")
+func (qs *queueSetImpl) isEmpty() bool {
+	return qs.numRequestsEnqueued == 0
+}
+
+// DequeueWithChannelAsMuchAsPossible runs a loop, as long as there
+// are non-empty queues and the number currently executing is less than the
+// assured concurrency value.  The body of the loop uses the fair queuing
+// technique to pick a queue, dequeue the request at the head of that
+// queue, increment the count of the number executing, and send `{true,
+// handleCompletion(that dequeued request)}` to the request's channel.
```

Comment:
This comment needs updating.  It is only `true` that is sent to the channel.

Reply:
Updated